### DISCUSSION TITLE
dynamically "map" select columns using with_only_columns

### DISCUSSION THREAD

**Discussion by timwoocker** [sqlalchemy/sqlalchemy#11963] [Usage Questions] [answered]
For an internal feature I need to dynamically change columns of a `Select` after creating it.
In my example, I need all columns which would map to python's `Decimal` to be exceplicitly casted to `Text`.
After finding `with_only_columns` in the docs, I tried that out like this (simplified example):

```python
columns = []
for col in config.select.columns:
    if col.type.python_type == Decimal:
        columns.append(Cast(func.trim_scale(col), Text()).label(col._label))
    else:
        columns.append(col)

inner_select = config.select.with_only_columns(*columns)
```

But the problem is already reproducable with this example expecting a `Select` in the variable `s`:

`s = s.with_only_columns(*s.columns)`

Compiling such a statement results in the usage of a subquery named `anon_1` which is bad because a subquery wouldn't be necessary here and it renames the resulting columns with the `anon_1` prefix:

```sql
SELECT anon_1.id, anon_1.data, CAST(trim_scale(anon_1.dec) AS TEXT) AS anon_1_dec 
FROM (SELECT new_table.id AS id, new_table.data AS data, new_table.dec AS dec 
FROM new_table) AS anon_1
```

I guess this behaviour somehow makes sense because I'm not passing the actual table's columns to `with_only_columns` but columns of a `Select` which therefore results in a subquery.
So I'm wondering if there is any good way of doing what I am trying to achieve without implicitly creating a subquery.

**Comment by zzzeek** [comment, marked as answer]
hi -

never use `s.columns` and that should be emitting a warning on 1.4.  this attribute is [deprecated](https://docs.sqlalchemy.org/en/20/changelog/migration_14.html#a-select-statement-is-no-longer-implicitly-considered-to-be-a-from-clause) in 1.4 and is totally gone in 2.0.

if you want the columns that are inside your select(), use [selected_columns](https://docs.sqlalchemy.org/en/20/core/selectable.html#sqlalchemy.sql.expression.SelectBase.selected_columns)

**Reply by timwoocker** [reply]
Using `selected_columns` was the answer, thank you!
But I didn't get a warning for using `columns` and I am on 2.0.35 but I was still able to access it

**Reply by timwoocker** [reply]
Also thank you for the great support, Michael!
I was going to "sponsor" the project via github but it won't let me for some reason. Is there a paypal address I can send a donation to?

**Reply by zzzeek** [reply]
looks like we didnt actually take it away from select() yet, here's that warning (it's in 1.4.x and 2.0.x):

```py
>>> from sqlalchemy import select, column
>>> s = select(column('q'))
>>> s.columns.q
<stdin>:1: SADeprecationWarning: The SelectBase.c and SelectBase.columns attributes are deprecated and will be removed in a future release; these attributes implicitly create a subquery that should be explicit.  Please call SelectBase.subquery() first in order to create a subquery, which then contains this attribute.  To access the columns that this SELECT object SELECTs from, use the SelectBase.selected_columns attribute. (deprecated since: 1.4)
<sqlalchemy.sql.elements.ColumnClause at 0x7feb5636bf80; q>

```

there's a paypal donate link at https://www.sqlalchemy.org/ for single shot donations, thanks !
